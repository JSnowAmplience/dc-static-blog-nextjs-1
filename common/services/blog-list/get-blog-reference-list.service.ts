import { DynamicContentDeliveryService } from '../dynamic-content-delivery.service';
import { BlogReferenceList, BlogPostReference } from '../../interfaces/blog-reference-list.interface';
import { defaultClientConfig } from '../dynamic-content-client-config';
import algoliasearch from 'algoliasearch';
import { isBlog } from '../../interfaces/blog.interface';

const INDEX_HITS_PER_PAGE = 1000;

export default async function getBlogReferenceList(
  blogListDeliveryKey: string,
  stagingEnvironment?: string
): Promise<BlogReferenceList> {
  if (
    !process.env.ALGOLIA_APPLICATION_ID ||
    !process.env.ALGOLIA_SEARCH_ONLY_KEY ||
    !process.env.ALGOLIA_PRODUCTION_INDEX_NAME
  ) {
    throw new Error('Missing Algolia configuration, please refer the readme');
  }
  // get the blogPost from the algolia index
  const client = algoliasearch(process.env.ALGOLIA_APPLICATION_ID, process.env.ALGOLIA_SEARCH_ONLY_KEY);
  const index = client.initIndex(process.env.ALGOLIA_PRODUCTION_INDEX_NAME);

  const results = await index.search<BlogPostReference>('', {
    attributesToRetrieve: ['deliveryKey'],
    attributesToHighlight: [],
    hitsPerPage: INDEX_HITS_PER_PAGE
  });

  // get the title and subTitle from DC
  const clientConfig = { ...defaultClientConfig, baseUrl: process.env.DYNAMIC_CONTENT_BASE_URL, stagingEnvironment };
  const deliveryClient = new DynamicContentDeliveryService(clientConfig);
  const blogContentItem = (await deliveryClient.getContentItemByKey(blogListDeliveryKey)).toJSON();
  if (!isBlog(blogContentItem)) {
    throw new Error(`Content item ${blogListDeliveryKey} is not a blog`);
  }

  return { ...blogContentItem, blogPosts: results.hits };
}
