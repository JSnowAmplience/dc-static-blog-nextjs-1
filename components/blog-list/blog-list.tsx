import BlogPost from '../../common/interfaces/blog-post.interface';
import BlogCard from '../blog-card/blog-card';
import theme from '../../common/styles/default/theme';

interface BlogListProps {
  blogPosts: BlogPost[];
}

const BlogList = ({ blogPosts = [] }: BlogListProps) => {
  const blogCards = blogPosts.map((blogPost: BlogPost, index: number) => <BlogCard key={index} blogPost={blogPost} />);

  return (
    <>
      <section>{blogCards.length ? blogCards : ''}</section>
      <style jsx>{`
        section {
          display: grid;
          grid-template-columns: repeat(3, 1fr);
          grid-gap: 55px 30px;
          grid-auto-rows: minmax(100px, auto);
          margin: auto;
          margin-top: 55px;
          max-width: ${theme.layout.widePageWidth};
        }

        @media (max-width: ${theme.layout.widePageWidth}) {
          section {
            grid-template-columns: repeat(2, 1fr);
            padding: 0 20px;
            margin-top: 0;
          }
        }

        @media (max-width: ${theme.layout.narrowPageWidth}) {
          section {
            display: block;
            padding: 0 45px;
          }
        }
      `}</style>
    </>
  );
};

export default BlogList;
