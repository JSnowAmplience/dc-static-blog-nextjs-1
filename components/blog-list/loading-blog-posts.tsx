import React, { ReactElement } from 'react';

const LoadingBlogPosts = (): ReactElement => {
  return (
    <>
      <div className="loading-blog-posts">
        <p>Loading...</p>
        <style jsx>{`
          .loading-blog-posts {
            padding-top: 60px;
            display: flex;
            justify-content: center;
            min-height: 500px;
          }
        `}</style>
      </div>
    </>
  );
};

export default LoadingBlogPosts;
