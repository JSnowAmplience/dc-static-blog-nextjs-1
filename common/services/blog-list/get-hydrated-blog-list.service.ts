import { BlogListData } from '../../interfaces/blog-list.interface';
import getBlogReferenceList from './get-blog-reference-list.service';
import getReferencedBlogPosts from '../blog-post/get-referenced-blog-posts.service';

export default async function getHydratedBlogList(
  blogListDeliveryKey: string,
  stagingEnvironment?: string
): Promise<BlogListData> {
  const { title, subTitle, blogPosts } = await getBlogReferenceList(blogListDeliveryKey, stagingEnvironment);
  const hydratedBlogPosts = await getReferencedBlogPosts(blogPosts, stagingEnvironment);
  return { title, subTitle, blogPosts: hydratedBlogPosts };
}
