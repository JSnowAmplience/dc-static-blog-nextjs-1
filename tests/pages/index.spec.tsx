/* eslint-env jest */
import renderer from 'react-test-renderer';
import Index from '../../pages/index';
import { MediaType } from '../../common/interfaces/media.interface';
import { BlogListData } from '../../common/interfaces/blog-list.interface';

const mockGetBlogReferenceList = jest.fn();
const mockGetBlogPost = jest.fn();

jest.mock('../../common/services/get-blog-reference-list.service', () => () => mockGetBlogReferenceList());
jest.mock('../../common/services/get-blog-post.service', () => () => mockGetBlogPost());

jest.mock('../../components/hero-banner/hero-banner', () => () => <div className="hero-banner-mock" />);
jest.mock('../../components/hero-card/hero-card', () => () => <div className="hero-card-mock" />);
jest.mock('../../components/blog-list/blog-list', () => () => <div className="blob-list-mock" />);

describe('Index', () => {
  let indexPropsFixture: BlogListData;

  beforeEach(() => {
    indexPropsFixture = {
      title: 'blog-test-title',
      subTitle: 'blog-test-sub-title',
      blogPosts: [
        {
          id: 'blog-post-id',
          title: 'blog-post-test',
          date: '2019-08-13',
          description: 'blog-post-description',
          authors: [],
          image: {
            image: {
              id: 'image-id',
              name: 'image-name',
              endpoint: 'image-endpoint',
              defaultHost: 'image-default-host',
              mediaType: MediaType.IMAGE
            },
            altText: ''
          },
          urlSlug: 'url-slug',
          tags: ['test-tag'],
          readTime: 15,
          content: []
        }
      ]
    };
  });
  test('renders index with content', async () => {
    const component = await renderer.create(<Index {...indexPropsFixture} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('getInitialProps returns content', async () => {
    const contentData = {
      title: 'blog-test-title',
      subTitle: 'blog-test-sub-title',
      blogPosts: [
        {
          id: '8d6943c7-6028-4fac-b45e-57fc63bd032a',
          _meta: {
            schema: 'http://bigcontent.io/cms/schema/v1/core#/definitions/content-reference'
          },
          contentType: 'https://schema.localhost.com/blog-post.json'
        }
      ]
    };
    mockGetBlogReferenceList.mockImplementation(() => {
      return contentData;
    });
    mockGetBlogPost.mockImplementation(() => indexPropsFixture.blogPosts[0]);
    const query = {};
    const result = await Index.getInitialProps({ query, pathname: '/' });

    expect(result).toEqual({ ...indexPropsFixture });
  });

  test('getInitialProps throws error when getContentItemById returns an error', async () => {
    mockGetBlogReferenceList.mockImplementation(() => {
      throw new Error();
    });
    const query = {};

    await expect(Index.getInitialProps({ query, pathname: '/' })).rejects.toThrowError(Error);
  });

  test('getInitialProps only returns resolved blog posts', async () => {
    const contentData = {
      title: 'blog-test-title',
      subTitle: 'blog-test-sub-title',
      blogPosts: [
        {
          id: '8d6943c7-6028-4fac-b45e-57fc63bd032a',
          _meta: {
            schema: 'http://bigcontent.io/cms/schema/v1/core#/definitions/content-reference'
          },
          contentType: 'https://schema.localhost.com/blog-post.json'
        },
        {
          id: '8d6943c7-6028-4fac-b45e-57fc63bd032b',
          _meta: {
            schema: 'http://bigcontent.io/cms/schema/v1/core#/definitions/content-reference'
          },
          contentType: 'https://schema.localhost.com/blog-post.json'
        }
      ]
    };
    mockGetBlogReferenceList.mockImplementation(() => {
      return contentData;
    });
    mockGetBlogPost
      .mockImplementationOnce(() => indexPropsFixture.blogPosts[0])
      .mockImplementationOnce(() => {
        throw new Error();
      });
    const query = {};
    const result = await Index.getInitialProps({ query, pathname: '/' });

    expect(result).toEqual({ ...indexPropsFixture });
  });
});
