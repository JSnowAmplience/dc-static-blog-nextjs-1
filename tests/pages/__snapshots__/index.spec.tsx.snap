// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index renders index with content 1`] = `
<div
  className="jsx-690136081 site-container"
>
  <section
    className="jsx-1853008300"
  >
    <header
      className="jsx-1853008300"
    >
      <a
        aria-label=""
        className="jsx-3354872520"
        href="/"
      >
        <img
          alt="Amplience logo"
          className="jsx-3359351483 logo"
          src="/static/images/logo-340w.png"
          srcSet="/static/images/logo-160w.png 1x, /static/images/logo-340w.png 2x"
        />
      </a>
      <h1
        className="jsx-1853008300"
      >
        PRODUCT BLOG
      </h1>
    </header>
  </section>
  <div
    className="jsx-690136081 site-content"
  >
    <main
      className="jsx-690136081"
    >
      <section
        className="jsx-2394770330"
      >
        <h1
          className="jsx-1446146482"
        >
          blog-test-title
        </h1>
        <h2
          className="jsx-1054465095"
        >
          blog-test-sub-title
        </h2>
      </section>
      <section
        className="jsx-1630848865"
      >
        <a
          aria-label="My First Blog Post"
          className="jsx-3354872520"
          href="/blog/my-first-blog"
        >
          <article
            className="jsx-1630848865"
          >
            <div
              className="jsx-1630848865 blog-card-image"
            >
              <span
                className="lazy-img"
                style={
                  Object {
                    "display": "inline-block",
                    "height": 0,
                    "width": 0,
                  }
                }
              />
            </div>
            <div
              className="jsx-1630848865 blog-card-content"
            >
              <h1
                className="jsx-1630848865"
              >
                My First Blog Post
              </h1>
              <div
                className="jsx-105857479 card-meta"
              >
                <div
                  className="jsx-105857479 authors"
                >
                  <span
                    className="jsx-105857479"
                  >
                    by
                  </span>
                  <span
                    className="jsx-105857479 author"
                  >
                    Sir Blog
                  </span>
                </div>
                <div
                  className="jsx-105857479 publish-data"
                >
                  <span
                    className="jsx-105857479 publish-date"
                  >
                    August 07, 2019
                  </span>
                  12 minutes
                </div>
              </div>
              <p
                className="jsx-1630848865"
              >
                Some text for the description that explains this blog post
              </p>
            </div>
          </article>
        </a>
      </section>
      <section
        className="jsx-2919536176"
      >
        <div
          className="jsx-143652140 no-published-blogs"
        >
          <p
            className="jsx-143652140"
          >
            No blogs published yet!
          </p>
        </div>
      </section>
    </main>
  </div>
  <footer
    className="jsx-27833366"
  >
    <section
      className="jsx-27833366"
    >
      <div
        className="jsx-27833366 footer-info"
      >
        <div
          className="jsx-27833366 footer-info-top"
        >
          <img
            alt="Amplience logo"
            className="jsx-3359351483 logo"
            src="/static/images/logo-340w.png"
            srcSet="/static/images/logo-160w.png 1x, /static/images/logo-340w.png 2x"
          />
          <div
            className="jsx-27833366 social-links"
          >
            <a
              className="jsx-3655793200"
              href="//twitter.com/amplience"
              rel="noreferrer"
              target="_new"
            >
              <img
                alt="Twitter logo"
                className="jsx-27833366"
                src="/static/images/social/ic_twitter.svg"
              />
            </a>
            <a
              className="jsx-3655793200"
              href="//www.linkedin.com/company/amplience/"
              rel="noreferrer"
              target="_new"
            >
              <img
                alt="Linkedin logo"
                className="jsx-27833366"
                src="/static/images/social/ic_linkedin.svg"
              />
            </a>
          </div>
        </div>
        <div
          className="jsx-27833366 copyright"
        >
          © Copyright Amplience 
          2019
        </div>
      </div>
      <div
        className="jsx-27833366 footer-links"
      >
        <section
          className="jsx-287747732 footer-links-section"
        >
          <h4
            className="jsx-287747732"
          >
            For Developers
          </h4>
          <ul
            className="jsx-287747732"
          >
            <li>
              <a
                className="jsx-3655793200"
                href="//github.com/amplience/"
                rel="noreferrer"
                target="_new"
              >
                GitHub
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//stackoverflow.com/questions/tagged/amplience-dynamic-content"
                rel="noreferrer"
                target="_new"
              >
                Stack Overflow
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//docs.amplience.net/integration/developergettingstarted.html"
                rel="noreferrer"
                target="_new"
              >
                Dynamic Content docs
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//docs.amplience.net/contenthub.html"
                rel="noreferrer"
                target="_new"
              >
                Content Hub docs
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//docs.amplience.net/integration/reference.html"
                rel="noreferrer"
                target="_new"
              >
                API reference
              </a>
            </li>
          </ul>
        </section>
        <section
          className="jsx-287747732 footer-links-section"
        >
          <h4
            className="jsx-287747732"
          >
            Company
          </h4>
          <ul
            className="jsx-287747732"
          >
            <li>
              <a
                className="jsx-3655793200"
                href="//amplience.com"
                rel="noreferrer"
                target="_new"
              >
                Website
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//amplience.com/careers"
                rel="noreferrer"
                target="_new"
              >
                Careers
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//amplience.com/privacy-cookie-policy"
                rel="noreferrer"
                target="_new"
              >
                Privacy policy
              </a>
            </li>
            <li>
              <a
                className="jsx-3655793200"
                href="//docs.amplience.net/releasenotes/readme.html"
                rel="noreferrer"
                target="_new"
              >
                What\`s new
              </a>
            </li>
          </ul>
        </section>
      </div>
    </section>
  </footer>
  <div
    className="cookieConsent "
    style={
      Object {
        "alignItems": "center",
        "background": "rgba(255, 255, 255, 0.9)",
        "bottom": "0",
        "boxShadow": "20px 20px 20px 10px #000",
        "color": "#333",
        "flexDirection": "column",
        "position": "fixed",
        "textAlign": "center",
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <div
      className="cookie-banner-content"
      style={
        Object {
          "padding": "20px 0px 10px",
        }
      }
    >
      We use cookies to give you the best possible experience on our website. By continuing to browse this site, you give consent for cookies to be used. For more details please read our
       
      <a
        href="https://amplience.com/privacy-cookie-policy/"
        target="_blank"
      >
        Cookie Policy
      </a>
      .
    </div>
    <button
      className="cookie-banner-btn"
      id=""
      onClick={[Function]}
      style={
        Object {
          "background": "#fff",
          "border": "3px solid #CCC",
          "borderRadius": "5px",
          "color": "#333",
          "marginBottom": "10px",
        }
      }
    >
      Accept Cookies
    </button>
  </div>
  <script
    async={true}
    src="https://www.google-analytics.com/analytics.js"
  />
  <script
    dangerouslySetInnerHTML={
      Object {
        "__html": "window.ga=window.ga||function(){(ga.q = ga.q || []).push(arguments)};ga.l=+new Date; ga('create', 'undefined', 'auto'); ga('send', 'pageview');",
      }
    }
  />
</div>
`;
